distributed_shampoo/fsdp_shampoo.py

```diff
@@ -0,0 +1,849 @@
+        exponent_override (int): exponent to use in Shampoo. (Default: 0)
+        exponent_multiplier (float): number to be multiplied to the numerator of the inverse root. (Default: 1.0)
+        use_nesterov (bool): uses Nesterov momentum (default: False)
+        use_bias_correction (bool): flag for using bias correction (Default: True)
+        use_decoupled_weight_decay (bool): Flag for using AdamW-style decoupled weight decay (Default: True)
+        preconditioner_dtype (torch.dtype): data type for preconditioner (Default: torch.float)
+        large_dim_method (LargeDimMethod): method for handling large scale tensors. (Default: LargeDimMethod.BLOCKING)
+        num_trainers_per_group (int): number of GPUs per distributed process group for distributed computation/memory.
+            If num_trainers_per_group = -1 is used, then defaults to using the LOCAL_WORLD_SIZE. (Default: -1)
+        use_merge_dims (bool): merge dimensions if possible while respecting max_preconditioner_dim. (Default: True)
+        grafting_type (GraftingType): selects grafting method. (Default: GraftingType.ADAGRAD)
+        grafting_epsilon (float): epsilon for grafting method. (Default: 1e-3)
+        grafting_beta2 (float): exponential moving average factor for grafting method. (Default: 1.0)
+        use_protected_eigh (bool): Flag for using two guards to prevent failures of torch.linalg.eigh. (Default: True)
+            1. Attempts to compute root inverse in preconditioner_dtype precision.
+            2. Attempts to recompute the eigendecomposition if using lower-precision fails.
+            3. Otherwise, re-uses previous inverse factor matrix when both root inverse computations fail.
+        use_dtensor (bool): use DTensor. Requires PyTorch 2 nightly. Otherwise, uses Tensor. (Default: True)
+        debug_mode (bool): debugging mode. Uses more memory to compute error to fp64 case. Must enable logging level to
+            DEBUG. (Default: False)
+
+    """
+
+    def __init__(
+        self,
+        params,
+        lr: float = 1e-2,
+        betas: Tuple[float, float] = (0.9, 1.0),
+        epsilon: float = 1e-12,
+        momentum: float = 0.0,
+        weight_decay: float = 0.0,
+        max_preconditioner_dim: int = 1024,
+        precondition_frequency: int = 1,
+        start_preconditioning_step: int = -1,
+        exponent_override: int = 0,
+        exponent_multiplier: float = 1.0,
+        use_nesterov: bool = False,
+        use_bias_correction: bool = True,
+        use_decoupled_weight_decay: bool = True,
+        preconditioner_dtype: torch.dtype = torch.float,
+        large_dim_method: LargeDimMethod = LargeDimMethod.BLOCKING,
+        num_trainers_per_group: int = -1,
+        use_merge_dims: bool = True,
+        grafting_type: GraftingType = GraftingType.ADAGRAD,
+        grafting_epsilon: float = 1e-3,
+        grafting_beta2: float = 1.0,
+        use_protected_eigh: bool = True,
+        use_dtensor: bool = True,
```

Comment:
Let's ignore using `DTensor` for now and set this to `False`, unless you think we need it? :)